### ISSUE TITLE
use std::optional in place of bool+returning by reference for YAML::convert::decode

### ISSUE THREAD

**Issue by cishisedis** [jbeder/yaml-cpp#696]
Since std::optional (specified by C++17) is now implemented by major compilers, it should make sense to relax the restriction of having a default constructor of the type to be decoded by adding an overload

- `std::optional<T> YAML::convert::decode(const Node&);`

for

- `bool YAML::convert::decode(const Node&, T&);`

if possible.

**Comment by david-antiteum** [comment]
IMHO it will better to wait (or use one of the unofficial implementations) for std::expected to change the api.

**Comment by jbeder** [comment]
Yeah, I think I agree with @david-antiteum, or at least, I don’t think optional is the right type here.

**Comment by SGSSGene** [comment]
Currently we are compatible with back to c++11, so using std::optional/std::variant/std::expected is not an option.
